### PR TITLE
Dynamic Concurrency support for WebJobs SDK Storage Extension

### PR DESCRIPTION
**Pull request by mathewc**
Implements the Dynamic Concurrency features in https://github.com/Azure/azure-functions-host/issues/7300.

Related PR for Dynamic Concurrency support in the Service Bus extension: https://github.com/Azure/azure-sdk-for-net/pull/23450.

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

File: eng/Packages.Data.props [hunk 1]
```diff
@@ -108,8 +108,8 @@
   <ItemGroup Condition="'$(IsClientLibrary)' == 'true' and $(MSBuildProjectName.StartsWith('Microsoft.'))">
     <PackageReference Update="CloudNative.CloudEvents" Version="2.0.0" />
     <PackageReference Update="CloudNative.CloudEvents.SystemTextJson" Version="2.0.0" />
-    <PackageReference Update="Microsoft.Azure.WebJobs" Version="3.0.27" />
-    <PackageReference Update="Microsoft.Azure.WebJobs.Sources" Version="3.0.19" />
+    <PackageReference Update="Microsoft.Azure.WebJobs" Version="3.0.30-11865" />
```
Comment (mathewc) [inline eng/Packages.Data.props:111]: This is a temporary reference to a preview package from our myget feed to allow this PR to compile. Before checkin, we'll reference a non-prerelease package from Nuget.

File: NuGet.Config [hunk 2]
```diff
@@ -10,6 +10,7 @@
       Used for azure-sdk-tools repo until issue https://github.com/Azure/azure-sdk-tools/issues/1329 is addressed
     -->
     <add key="azure-sdk-tools" value="https://azuresdkartifacts.blob.core.windows.net/azure-sdk-tools/index.json" />
+    <add key="azure_app_service" value="https://www.myget.org/F/azure-appservice/api/v2" />
```
Comment (mathewc) [inline NuGet.Config:13]: Temporary - won't be checked in

Comment (weshaggard) [reply NuGet.Config:13]: Thanks was about to ask about this. What is the plan to consume the correct packages?

Comment (mathewc) [reply NuGet.Config:13]: We'll be publishing the non-prerelease packages to nuget.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs [hunk 3]
```diff
@@ -281,6 +290,38 @@ public void Notify()
...
+        internal int GetMessageReceiveCount(out TaskSeriesCommandResult delay)
+        {
+            delay = default(TaskSeriesCommandResult);
+            int numMessagesToReceive = _queueProcessor.QueuesOptions.BatchSize;
+
+            if (_concurrencyManager != null && _concurrencyManager.Enabled)
+            {
+                // When not using DC, WaitForNewBatchThreshold cleans up tasks as they complete.
+                // When using DC we need to also do cleanup of completed tasks, particularly
+                // before we use the count of pending tasks in the calculation below.
```
Comment (pakrym) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs:302]: Is having correct count the only goal? If so can we do `_processing.Count(p => !p.IsCompleted)` and keep the cleanup in a single place?

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs [hunk 4]
```diff
@@ -297,10 +338,26 @@ private TaskSeriesCommandResult CreateBackoffResult()
...
+            Task aggregateTask = Task.WhenAny(_stopWaitingTaskSource.Task, Task.Delay(delay));
+
+            return new TaskSeriesCommandResult(wait: aggregateTask);
+        }
+
         private TaskSeriesCommandResult CreateSucceededResult()
         {
-            Task wait = WaitForNewBatchThreshold();
-            return new TaskSeriesCommandResult(wait);
+            if (_concurrencyManager != null && _concurrencyManager.Enabled)
```
Comment (pakrym) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs:350]: In the listener factory we have a null check `_concurrencyManager = concurrencyManager ?? throw new ArgumentNullException(nameof(concurrencyManager));`

Considering that, do we need all the `_concurrencyManager != null` in this file?

**Review by pakrym — APPROVED** [review]
LGTM with some minor nits, I'd like @josh-love to review as well.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs (hunk 3 above)
Comment (mathewc) [reply sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs:302]: It's not just to have a correct count, it's also to remove completed tasks. When DC is enabled, the other cleanup logic [here](https://github.com/Azure/azure-sdk-for-net/blob/4162f6fa2445b2127468b9cfd080f01c9da88eba/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs#L306) isn't running.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs (hunk 4 above)
Comment (mathewc) [reply sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs:350]: There are a few paths into the QueueListener ctor. What I'll do is add a null check in the QueueListener ctor (ConcurrencyManager is an optional param currently), and make sure in all cases it'll be non-null. Then I'll remove the other null checks in this file.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Scenario.Tests/tests/AzureStorageEndToEndTests.cs [hunk 5]
```diff
@@ -160,7 +167,144 @@ public async Task AzureStorageEndToEndFast()
...
+                {
+                    services.AddSingleton<INameResolver>(_resolver);
+                })
+                .ConfigureLogging((context, b) =>
+                {
+                    b.SetMinimumLevel(LogLevel.Debug);
+                })
+                .Build();
+
+            string sharedListenerId = "SharedBlobQueueListener";
```
Comment (mathewc) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Scenario.Tests/tests/AzureStorageEndToEndTests.cs:276]: I've added an E2E test for blobs. This shows the shared listener pattern - the fact that internally there's a virtual queue trigger function "SharedBlobQueueListener".

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs [hunk 6]
```diff
@@ -199,9 +203,16 @@ public async Task<TaskSeriesCommandResult> ExecuteAsync(CancellationToken cancel
 
                     if (_queueExists.Value)
                     {
-                        sw = Stopwatch.StartNew();
+                        int numMessagesToReceive = GetMessageReceiveCount();
+                        if (numMessagesToReceive == 0)
+                        {
+                            // if no messages can be fetched because we're at our limit or because
+                            // throttles are enabled, wait a short period before checking again
+                            return CreateDelayResult(TimeSpan.FromMilliseconds(50));
```
Comment (kasobol-msft) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs:211]: should this be configurable ?

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs [hunk 7]
```diff
@@ -281,6 +292,31 @@ public void Notify()
             }
         }
 
+        internal int GetMessageReceiveCount()
+        {
+            int numMessagesToReceive = _queueProcessor.QueuesOptions.BatchSize;
+
+            if (_concurrencyManager != null && _concurrencyManager.Enabled)
```
Comment (kasobol-msft) [inline sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs:299]: How user knows that concurrency manager is used? It looks like BatchSize setting will be ignored in that case. Should we emit a log somewhere in init phase that this is going to happen?

**Review by kasobol-msft — COMMENTED** [review]
LGTM. I suggest to run "live tests" when dependencies are available in nuget before merging.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs [hunk 8]
```diff
@@ -281,6 +290,38 @@ public void Notify()
...
+                // When DynamicConcurrency is enabled, we determine the number of messages to pull
+                // based on the current degree of parallelism for this function and our pending invocation count.
+                var concurrencyStatus = _concurrencyManager.GetStatus(_functionId);
+                int availableInvocationCount = concurrencyStatus.GetAvailableInvocationCount(_processing.Count);
+                numMessagesToReceive = Math.Min(availableInvocationCount, QueuesOptions.MaxBatchSize);
+                if (numMessagesToReceive == 0)
+                {
+                    // if we're not healthy or we're at our limit, we'll wait
+                    // a bit before checking again
+                    delay = CreateDelayResult(TimeSpan.FromSeconds(1));
```
[the remark this answers is not in the document]
Comment (paulbatum) [reply sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs]: Thats definitely an improvement but 50ms stalls are still a problem - its like having a random thread.sleep(50) in your hot path. I think there is a design flaw here if the system cannot sustain constant throughput when this feature is enabled.

[the remark this answers is not in the document]
Comment (mathewc) [reply sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs]: Ok, I've removed the delay completely, which will be OK in tandem with changes I'm making in https://github.com/Azure/azure-webjobs-sdk/pull/2758 to reduce logging.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs (hunk 6 above)
Comment (mathewc) [reply sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs]: The delay has been removed.

File: sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs (hunk 7 above)
Comment (mathewc) [reply sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Common/src/Shared/Queues/QueueListener.cs:299]: On host startup, we log ConcurrencyOptions at the host level which will show DC is enabled globally. Additionally the user will see periodic concurrency logs from DC for each of their functions using DC. So I think we're covered.

**Review by paulbatum — APPROVED** [review]

**Pull request merged**